source/agora/node/Ledger.d

```diff
@@ -1370,6 +1392,31 @@ public class ValidatingLedger : Ledger
             clock, block_timestamp_tolerance, onAcceptedBlock);
     }
 
+    // dynamic array to keep track of blocks we are externalizing so can allow
+    //  less signatures than majority when validating
+    // TODO: We need to clear out old entries to reduce memory footprint
+    private Height[] externalizing;
+
+    public void addHeightAsExternalizing (Height height) @safe nothrow
+    {
+        externalizing ~= height;
+    }
+
+    /***************************************************************************
+
+        Used to indicate if the block we are validating has been triggered to
+            externalize by the SCP event EXTERNALIZE.
+
+        Params:
+            height = ignored here but used in ValidatingLedger function override
+
+    ***************************************************************************/
+
+    protected override bool isExternalizing (in Height height) @safe nothrow
+    {
+        return externalizing.canFind(height);
+    }
+
```

Comment:
Well, first, the Ledger is not guaranteed to be a `ValidatingLedger`. So having a `isExternalizing` in the base class and overriding it here is still leaking abstraction. Overriding `validateBlockSignature` seems more sensible.

Also, even if we're going with this, is there a situation in which we would sign a block without signing intermediate blocks ? As in, do we ever sign block `N + 1` before block `N` ?

Reply:
> Well, first, the Ledger is not guaranteed to be a `ValidatingLedger`. So having a `isExternalizing` in the base class and overriding it here is still leaking abstraction. Overriding `validateBlockSignature` seems more sensible.

I'll fix the leaking abstraction.  did consider overriding `validateBlockSignature` but I think it ended up with some duplication. Maybe a better fit is to have a `checkMajority` function which is overridden in the `ValidatingLedger`.

Reply:
> Also, even if we're going with this, is there a situation in which we would sign a block without signing intermediate blocks ? As in, do we ever sign block `N + 1` before block `N` ?

The problem is not if we have signed the block, but if enough of the others have signed and we have received them. If we `externalize` blocks quickly we can be waiting on more than one block to have more than half signatures added.

Reply:
> If we externalize blocks quickly we can be waiting on more than one block to have more than half signatures added.

Do we externalize blocks on top of blocks for which we don't have a majority signature ?

Reply:
We do if we are a `Validator` who is `externalizing` blocks due to `SCP` `EXTERNALIZE` events. Other nodes will not `externalize` blocks until they have the required majority of signatures. I think this is ok if we are relying on the fact that `SCP` `EXTERNALIZE` event can only happen once per block height. Also, I am not sure we can avoid this if we want regular block intervals.